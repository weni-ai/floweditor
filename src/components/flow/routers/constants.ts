import i18n from 'config/i18n';

export const WAIT_LABEL = i18n.t(
  'forms.message_label',
  'If the message response...',
);
export const EXPRESSION_LABEL = i18n.t(
  'forms.expression_label',
  'If the expression...',
);
export const GROUP_LABEL = i18n.t(
  'forms.split_by_groups',
  "Select the groups you'd like to split by below",
);

export const OPERAND_LOCALIZATION_DESC = i18n.t(
  'forms.localize_rules',
  'Sometimes languages need special rules to route things properly. If a translation is not provided, the original rule will be used.',
);

// We intentionally don't localize these names
export enum DefaultExitNames {
  All_Responses = 'All Responses',
  No_Response = 'No Response',
  Any_Value = 'Any Value',
  Other = 'Other',
}
