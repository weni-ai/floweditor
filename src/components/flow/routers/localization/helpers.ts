import { Types } from 'config/interfaces';
import { getType } from 'config/typeConfigs';
import { Case, Category, SwitchRouter } from 'flowTypes';
import { LocalizedObject } from 'services/Localization';
import { RenderNode } from 'store/flowContext';
import { NodeEditorSettings } from 'store/nodeEditor';

export enum LocalizedType {
  Category,
  Case,
}

export const getOriginalCase = (
  nodeSettings: NodeEditorSettings,
  uuid: string,
) => {
  const cases = (nodeSettings.originalNode.node.router as SwitchRouter).cases;
  return cases.find((item: any) => item.uuid === uuid);
};

export const getOriginalCategory = (
  nodeSettings: NodeEditorSettings,
  uuid: string,
) => {
  const items = nodeSettings.originalNode.node.router.categories;
  return items.find((item: any) => item.uuid === uuid);
};

export const hasLocalizableCases = (renderNode: RenderNode) => {
  const type = getType(renderNode);
  return [
    Types.wait_for_response,
    Types.split_by_expression,
    Types.smart_wait_for_response,
    Types.automatic_classify,
  ].includes(type);
};

export const getLocalizedObjects = (
  nodeSettings: NodeEditorSettings,
  localizedType: LocalizedType,
): Category[] | Case[] => {
  const filtered: any = [];

  let items: Category[] | Case[] =
    nodeSettings.originalNode.node.router.categories;
  if (localizedType === LocalizedType.Case) {
    if (hasLocalizableCases(nodeSettings.originalNode)) {
      items = (nodeSettings.originalNode.node.router as SwitchRouter).cases;
    } else {
      items = [];
    }
  }

  for (const original of items) {
    const [localized] = nodeSettings.localizations.filter(
      (localizedObject: LocalizedObject) =>
        localizedObject.getObject().uuid === original.uuid,
    );

    if (localized && localized.isLocalized()) {
      filtered.push(localized.getObject());
    } else {
      filtered.push({ uuid: original.uuid });
    }
  }
  return filtered;
};
