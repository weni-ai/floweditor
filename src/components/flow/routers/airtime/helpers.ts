import {
  AirtimeRouterFormState,
  AirtimeTransferEntry,
} from 'components/flow/routers/airtime/AirtimeRouterForm';
import { createWebhookBasedNode } from 'components/flow/routers/helpers';
import { Types } from 'config/interfaces';
import { TransferAirtime } from 'flowTypes';
import { RenderNode } from 'store/flowContext';
import { NodeEditorSettings } from 'store/nodeEditor';
import { createUUID } from 'utils';

export const nodeToState = (
  settings: NodeEditorSettings,
): AirtimeRouterFormState => {
  const originalAction = getOriginalAction(settings);
  let resultName = { value: 'Result' };
  let valid = false;

  const amounts: AirtimeTransferEntry[] = [];
  if (originalAction && originalAction.type === Types.transfer_airtime) {
    Object.keys(originalAction.amounts).forEach((key: string) => {
      amounts.push({
        value: { code: key, amount: '' + originalAction.amounts[key] },
      });
    });
    resultName = { value: originalAction.result_name };
    valid = true;
  }

  return {
    valid,
    amounts,
    resultName,
  };
};

export const stateToNode = (
  settings: NodeEditorSettings,
  state: AirtimeRouterFormState,
): RenderNode => {
  let uuid = createUUID();
  const originalAction = getOriginalAction(settings);
  if (originalAction) {
    uuid = originalAction.uuid;
  }

  const amounts = {};
  state.amounts.forEach((entry: AirtimeTransferEntry) => {
    if (entry.value.amount.trim().length > 0) {
      (amounts as any)[entry.value.code] = Number(entry.value.amount);
    }
  });

  const newAction: TransferAirtime = {
    uuid,
    type: Types.transfer_airtime,
    amounts,
    result_name: state.resultName.value,
  };

  return createWebhookBasedNode(newAction, settings.originalNode, true);
};

export const getOriginalAction = (
  settings: NodeEditorSettings,
): TransferAirtime => {
  const action =
    settings.originalAction ||
    (settings.originalNode.node.actions.length > 0 &&
      settings.originalNode.node.actions[0]);

  if (action.type === Types.transfer_airtime) {
    return action as TransferAirtime;
  }
};
