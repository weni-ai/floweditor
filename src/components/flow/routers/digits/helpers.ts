import { CaseProps } from 'components/flow/routers/caselist/CaseList';
import {
  createCaseProps,
  createRenderNode,
  hasCases,
  resolveRoutes,
} from 'components/flow/routers/helpers';
import { Types } from 'config/interfaces';
import { getType } from 'config/typeConfigs';
import {
  HintTypes,
  Router,
  RouterTypes,
  SwitchRouter,
  WaitTypes,
} from 'flowTypes';
import { RenderNode } from 'store/flowContext';
import { NodeEditorSettings, StringEntry } from 'store/nodeEditor';

import { DigitsRouterFormState } from './DigitsRouterForm';
import { DIGITS_OPERAND } from 'components/nodeeditor/constants';

export const nodeToState = (
  settings: NodeEditorSettings,
): DigitsRouterFormState => {
  let initialCases: CaseProps[] = [];

  // TODO: work out an incremental result name
  let resultName: StringEntry = { value: '' };

  if (getType(settings.originalNode) === Types.wait_for_digits) {
    const router = settings.originalNode.node.router as SwitchRouter;
    if (router) {
      if (hasCases(settings.originalNode.node)) {
        initialCases = createCaseProps(router.cases, settings.originalNode);
      }

      resultName = { value: router.result_name || '' };
    }
  }

  return {
    cases: initialCases,
    resultName,
    valid: true,
  };
};

export const stateToNode = (
  settings: NodeEditorSettings,
  state: DigitsRouterFormState,
): RenderNode => {
  const {
    cases,
    exits,
    categories,
    defaultCategory: defaultExit,
    caseConfig,
  } = resolveRoutes(state.cases, false, settings.originalNode.node);

  const optionalRouter: Pick<Router, 'result_name'> = {};
  if (state.resultName.value) {
    optionalRouter.result_name = state.resultName.value;
  }

  const router: SwitchRouter = {
    type: RouterTypes.switch,
    default_category_uuid: defaultExit,
    categories,
    cases,
    operand: DIGITS_OPERAND,
    wait: { type: WaitTypes.msg, hint: { type: HintTypes.digits } },
    ...optionalRouter,
  };

  const newRenderNode = createRenderNode(
    settings.originalNode.node.uuid,
    router,
    exits,
    Types.wait_for_response,
    [],

    { cases: caseConfig },
  );

  return newRenderNode;
};
