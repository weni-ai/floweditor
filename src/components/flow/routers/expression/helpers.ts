import { CaseProps } from 'components/flow/routers/caselist/CaseList';
import { ExpressionRouterFormState } from 'components/flow/routers/expression/ExpressionRouterForm';
import {
  createCaseProps,
  createRenderNode,
  hasCases,
  resolveRoutes,
} from 'components/flow/routers/helpers';
import { DEFAULT_OPERAND } from 'components/nodeeditor/constants';
import { Types } from 'config/interfaces';
import { getType } from 'config/typeConfigs';
import { Router, RouterTypes, SwitchRouter } from 'flowTypes';
import { RenderNode } from 'store/flowContext';
import { NodeEditorSettings, StringEntry } from 'store/nodeEditor';

export const nodeToState = (
  settings: NodeEditorSettings,
): ExpressionRouterFormState => {
  let initialCases: CaseProps[] = [];

  // TODO: work out an incremental result name
  let resultName: StringEntry = { value: '' };

  let operand = DEFAULT_OPERAND;

  if (
    settings.originalNode &&
    getType(settings.originalNode) === Types.split_by_expression
  ) {
    const router = settings.originalNode.node.router as SwitchRouter;
    if (router) {
      if (hasCases(settings.originalNode.node)) {
        initialCases = createCaseProps(router.cases, settings.originalNode);
      }

      resultName = { value: router.result_name || '' };
    }

    operand = router.operand;
  }

  return {
    cases: initialCases,
    resultName,
    operand: { value: operand },
    valid: true,
  };
};

export const stateToNode = (
  settings: NodeEditorSettings,
  state: ExpressionRouterFormState,
): RenderNode => {
  const {
    cases,
    exits,
    defaultCategory: defaultExit,
    caseConfig,
    categories,
  } = resolveRoutes(state.cases, false, settings.originalNode.node);

  const optionalRouter: Pick<Router, 'result_name'> = {};
  if (state.resultName.value) {
    optionalRouter.result_name = state.resultName.value;
  }

  const router: SwitchRouter = {
    type: RouterTypes.switch,
    default_category_uuid: defaultExit,
    categories,
    cases,
    operand: state.operand.value,
    ...optionalRouter,
  };

  const newRenderNode = createRenderNode(
    settings.originalNode.node.uuid,
    router,
    exits,
    Types.split_by_expression,
    [],
    { cases: caseConfig },
  );

  return newRenderNode;
};
