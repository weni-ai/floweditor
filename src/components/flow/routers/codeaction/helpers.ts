import { createRenderNode } from 'components/flow/routers/helpers';
import {
  Category,
  Exit,
  RouterTypes,
  SwitchRouter,
  WebhookExitNames,
} from 'flowTypes';
import { Types } from 'config/interfaces';
import { RunCodeAction } from 'flowTypes';
import { RenderNode } from 'store/flowContext';
import { NodeEditorSettings, FormEntry } from 'store/nodeEditor';
import { createUUID, snakify } from 'utils';
import { CodeActionRouterFormState } from 'components/flow/routers/codeaction/CodeActionRouterForm';
import { StringEntry } from '../../../../store/nodeEditor';

export const getOriginalAction = (
  settings: NodeEditorSettings,
): RunCodeAction => {
  const action =
    settings.originalAction ||
    (settings.originalNode.node.actions.length > 0 &&
      settings.originalNode.node.actions[0]);

  if (action.type === Types.run_code_action) {
    return action as RunCodeAction;
  }
};

export const nodeToState = (
  settings: NodeEditorSettings,
  initialCodeAction: any,
): CodeActionRouterFormState => {
  const originalAction = getOriginalAction(settings);
  let resultName: StringEntry = { value: 'Result' };
  let codeAction: FormEntry = {
    value: { id: 0, name: 'Select a Code Action' },
  };

  if (originalAction && originalAction.type === Types.run_code_action) {
    codeAction = { value: originalAction.codeaction };
    resultName = { value: originalAction.result_name };
  } else {
    if (initialCodeAction) {
      codeAction = initialCodeAction;
    }
  }

  const state: CodeActionRouterFormState = {
    resultName,
    codeAction,
    valid: true,
  };
  return state;
};

export const stateToNode = (
  settings: NodeEditorSettings,
  state: CodeActionRouterFormState,
): RenderNode => {
  let uuid = createUUID();
  const originalAction = getOriginalAction(settings);
  if (originalAction) {
    uuid = originalAction.uuid;
  }
  let categories: Category[] = [];
  let exits: Exit[] = [];

  const newAction: RunCodeAction = {
    uuid,
    type: Types.run_code_action,
    codeaction: {
      id: state.codeAction.value.id,
      name: state.codeAction.value.name,
    },
    result_name: state.resultName.value,
  };

  const originalNode = settings.originalNode;
  if (
    originalNode &&
    originalNode.node.actions.length === 1 &&
    originalNode.node.actions[0].type === newAction.type
  ) {
    exits = originalNode.node.exits;
    categories = originalNode.node.router.categories;
  } else {
    exits = [
      {
        uuid: createUUID(),
        destination_uuid: null,
      },
      {
        uuid: createUUID(),
        destination_uuid: null,
      },
    ];

    categories = [
      {
        uuid: createUUID(),
        name: WebhookExitNames.Success,
        exit_uuid: exits[0].uuid,
      },
      {
        uuid: createUUID(),
        name: WebhookExitNames.Failure,
        exit_uuid: exits[1].uuid,
      },
    ];
  }

  let operand = '@results.' + snakify(newAction.result_name);

  const router: SwitchRouter = {
    type: RouterTypes.switch,
    default_category_uuid: categories[0].uuid,
    cases: [],
    categories,
    operand,
  };

  const newRenderNode = createRenderNode(
    settings.originalNode.node.uuid,
    router,
    exits,
    Types.run_code_action,
    [newAction],
  );

  return newRenderNode;
};
