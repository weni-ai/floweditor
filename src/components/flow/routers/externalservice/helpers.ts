import { createWebhookBasedNode } from 'components/flow/routers/helpers';
import { ServiceCall, Types } from 'config/interfaces';
import { getType } from 'config/typeConfigs';
import { CallExternalService } from 'flowTypes';
import { AssetMap, RenderNode } from 'store/flowContext';
import { NodeEditorSettings, FormEntry } from 'store/nodeEditor';
import { createUUID } from 'utils';
import { ExternalServiceRouterFormState } from 'components/flow/routers/externalservice/ExternalServiceRouterForm';

export const getOriginalAction = (
  settings: NodeEditorSettings,
): CallExternalService => {
  const action =
    settings.originalAction ||
    (settings.originalNode.node.actions.length > 0 &&
      settings.originalNode.node.actions[0]);

  if (action.type === Types.call_external_service) {
    return action as CallExternalService;
  }
};

export const nodeToState = (
  settings: NodeEditorSettings,
  existingServices: AssetMap,
  initialExternalService: any,
): ExternalServiceRouterFormState => {
  let resultName = { value: 'Result' };

  let externalService: FormEntry = null;
  let initialCall: FormEntry = null;
  let initialParams: any[] = [];

  if (
    settings.originalNode &&
    getType(settings.originalNode) === Types.split_by_external_service
  ) {
    const action = getOriginalAction(settings) as CallExternalService;
    const service = existingServices[action.external_service.uuid];

    if (service) {
      externalService = {
        value: {
          ...service.content,
          name: service.name,
          external_service_type: service.type,
          uuid: service.id,
        },
      };
    } else {
      externalService = { value: action.external_service };
    }

    const existingCallParams =
      externalService.value.actions &&
      externalService.value.actions.find((a: ServiceCall) => {
        return a.name === action.call.name;
      }).params;

    initialCall = {
      value: existingCallParams
        ? { ...action.call, params: existingCallParams }
        : action.call,
    };
    initialParams = action.params;
    resultName = { value: action.result_name };
  } else {
    externalService = initialExternalService
      ? {
          value: {
            uuid: initialExternalService.id,
            name: initialExternalService.name,
            type: initialExternalService.type,
            actions: initialExternalService.content.actions,
          },
        }
      : { value: null };
  }

  const state: ExternalServiceRouterFormState = {
    externalService,
    call: initialCall,
    resultName,
    params: { value: initialParams },
    valid: true,
  };

  return state;
};

export const stateToNode = (
  settings: NodeEditorSettings,
  state: ExternalServiceRouterFormState,
): RenderNode => {
  let uuid = createUUID();
  const originalAction = getOriginalAction(settings);
  if (originalAction) {
    uuid = originalAction.uuid;
  }

  const validParams = state.params.value.filter(
    (param: any) =>
      param.valid === true &&
      param.data.value &&
      param.data.validationFailures.length === 0,
  );

  const newAction: CallExternalService = {
    uuid,
    type: Types.call_external_service,
    external_service: {
      uuid: state.externalService.value.uuid,
      name: state.externalService.value.name,
      external_service_type:
        state.externalService.value.external_service_type ||
        state.externalService.value.type,
      actions: state.externalService.value.actions,
    },
    call: state.call.value,
    params: validParams,
    result_name: state.resultName.value,
  };

  return createWebhookBasedNode(newAction, settings.originalNode, true);
};
