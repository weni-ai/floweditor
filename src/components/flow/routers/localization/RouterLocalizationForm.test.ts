import { LocalizationFormProps } from 'components/flow/props';
import RouterLocalizationForm from 'components/flow/routers/localization/RouterLocalizationForm';
import { DEFAULT_OPERAND } from 'components/nodeeditor/constants';
import { Operators } from 'config/interfaces';
import { RouterTypes, SwitchRouter, Exit, Category } from 'flowTypes';
import { getLocalizations } from 'store/helpers';
import { composeComponentTestUtils, mock } from 'testUtils';
import {
  createRenderNode,
  Spanish,
  createCategories,
  createMatchRouter,
} from 'testUtils/assetCreators';
import * as utils from 'utils';
import { getSmartOrSwitchRouter } from 'components/flow/routers/helpers';

mock(utils, 'createUUID', utils.seededUUIDs());

const responseRenderNode = createMatchRouter(['Red', 'Blue']);

const router = getSmartOrSwitchRouter(responseRenderNode.node);

const redCase = router.cases[0];
const otherCategory = router.categories[router.categories.length - 1];

const localizations = getLocalizations(responseRenderNode.node, null, Spanish, {
  [redCase.uuid]: { arguments: ['rojo, r'] },

  [otherCategory.uuid]: { name: ['Otro'] },
});

const baseProps: LocalizationFormProps = {
  language: Spanish,
  updateLocalizations: jest.fn(),
  onClose: jest.fn(),
  nodeSettings: {
    originalNode: responseRenderNode,
    originalAction: null,
    localizations,
  },
};

const { setup } = composeComponentTestUtils<LocalizationFormProps>(
  RouterLocalizationForm,
  baseProps,
);

describe(RouterLocalizationForm.name, () => {
  it('should render', () => {
    const { wrapper } = setup(true);
    expect(wrapper).toMatchSnapshot();
  });

  it('initializes', () => {
    const { wrapper } = setup(true);
    expect(wrapper).toMatchSnapshot();
  });

  describe('updates', () => {
    it('should save changes', () => {
      const { instance, props } = setup(true);

      instance.handleUpdateCategoryName(
        responseRenderNode.node.router.categories[0],
        'Roooojo!',
      );
      instance.handleUpdateCaseArgument(
        (responseRenderNode.node.router as SwitchRouter).cases[0],
        'Red, r, rolo, maroon',
      );
      expect(instance.state).toMatchSnapshot();

      instance.handleSave();
      expect(props.updateLocalizations).toHaveBeenCalled();
      expect(props.updateLocalizations).toMatchCallSnapshot();
    });
  });
});
