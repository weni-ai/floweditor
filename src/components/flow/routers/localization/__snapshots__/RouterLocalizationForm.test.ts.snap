// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`RouterLocalizationForm > initializes 1`] = `
<Dialog
  buttons={
    {
      "primary": {
        "name": "Confirm",
        "onClick": [Function],
      },
      "secondary": {
        "name": "Cancel",
        "onClick": [Function],
      },
    }
  }
  headerClass="wait_for_response"
  tabs={
    [
      {
        "body": <React.Fragment>
          <p
            data-spec="instructions"
          >
            Sometimes languages need special rules to route things properly. If a translation is not provided, the original rule will be used.
          </p>
          <div
            className="_translating_list_container_c108c5 "
            tabIndex={0}
          >
            <div
              className="_translating_item_list_c108c5"
            >
              <div
                className="_translating_operator_container_c108c5"
                data-spec="operator-field"
              >
                <div
                  className="_translating_operator_c108c5"
                  data-spec="verbose-name"
                >
                  has any of the words
                </div>
                <div
                  className="_translating_from_c108c5"
                  data-spec="argument-to-translate"
                >
                  red
                </div>
                <div
                  className="_translating_to_c108c5"
                >
                  <TextInputElement
                    data-spec="localize-case"
                    entry={
                      {
                        "value": "rojo, r",
                      }
                    }
                    name="061fc171-8b79-4636-b892-bd0ea5aa9b42"
                    onChange={[Function]}
                    placeholder="Spanish Translation"
                    showLabel={false}
                  />
                </div>
              </div>
              <div
                className="_translating_operator_container_c108c5"
                data-spec="operator-field"
              >
                <div
                  className="_translating_operator_c108c5"
                  data-spec="verbose-name"
                >
                  has any of the words
                </div>
                <div
                  className="_translating_from_c108c5"
                  data-spec="argument-to-translate"
                >
                  blue
                </div>
                <div
                  className="_translating_to_c108c5"
                >
                  <TextInputElement
                    data-spec="localize-case"
                    entry={
                      {
                        "value": "",
                      }
                    }
                    name="763e4844-3e1b-407a-a1b5-5fdfcd308b41"
                    onChange={[Function]}
                    placeholder="Spanish Translation"
                    showLabel={false}
                  />
                </div>
              </div>
            </div>
          </div>
        </React.Fragment>,
        "name": "Rule Translations",
      },
    ]
  }
  title="Spanish Category Names"
>
  <p
    data-spec="instructions"
  >
    When category names are referenced later in the flow, the appropriate language for the category will be used. If no translation is provided, the original text will be used.
  </p>
  <div
    className="_translating_list_container_c108c5 "
    tabIndex={0}
  >
    <div
      className="_translating_item_list_c108c5"
    >
      <div
        className="_translating_category_c108c5"
        key="1e47a1e1-3c67-4df5-adf1-da542c789adb"
      >
        <div
          className="_translating_from_c108c5"
          data-spec="category-name"
        >
          Red
        </div>
        <div
          className="_translating_to_c108c5"
        >
          <TextInputElement
            data-spec="localize-category"
            entry={
              {
                "value": "",
              }
            }
            name=""
            onChange={[Function]}
            placeholder="Spanish Translation"
            showLabel={false}
          />
        </div>
      </div>
      <div
        className="_translating_category_c108c5"
        key="2dc85899-0577-4f1b-a620-f12094e34b5e"
      >
        <div
          className="_translating_from_c108c5"
          data-spec="category-name"
        >
          Blue
        </div>
        <div
          className="_translating_to_c108c5"
        >
          <TextInputElement
            data-spec="localize-category"
            entry={
              {
                "value": "",
              }
            }
            name=""
            onChange={[Function]}
            placeholder="Spanish Translation"
            showLabel={false}
          />
        </div>
      </div>
      <div
        className="_translating_category_c108c5"
        key="70ac6ea6-803a-4c33-81c7-dc26803c313f"
      >
        <div
          className="_translating_from_c108c5"
          data-spec="category-name"
        >
          Other
        </div>
        <div
          className="_translating_to_c108c5"
        >
          <TextInputElement
            data-spec="localize-category"
            entry={
              {
                "value": "Otro",
              }
            }
            name="Otro"
            onChange={[Function]}
            placeholder="Spanish Translation"
            showLabel={false}
          />
        </div>
      </div>
    </div>
  </div>
</Dialog>
`;

exports[`RouterLocalizationForm > should render 1`] = `
<Dialog
  buttons={
    {
      "primary": {
        "name": "Confirm",
        "onClick": [Function],
      },
      "secondary": {
        "name": "Cancel",
        "onClick": [Function],
      },
    }
  }
  headerClass="wait_for_response"
  tabs={
    [
      {
        "body": <React.Fragment>
          <p
            data-spec="instructions"
          >
            Sometimes languages need special rules to route things properly. If a translation is not provided, the original rule will be used.
          </p>
          <div
            className="_translating_list_container_c108c5 "
            tabIndex={0}
          >
            <div
              className="_translating_item_list_c108c5"
            >
              <div
                className="_translating_operator_container_c108c5"
                data-spec="operator-field"
              >
                <div
                  className="_translating_operator_c108c5"
                  data-spec="verbose-name"
                >
                  has any of the words
                </div>
                <div
                  className="_translating_from_c108c5"
                  data-spec="argument-to-translate"
                >
                  red
                </div>
                <div
                  className="_translating_to_c108c5"
                >
                  <TextInputElement
                    data-spec="localize-case"
                    entry={
                      {
                        "value": "rojo, r",
                      }
                    }
                    name="061fc171-8b79-4636-b892-bd0ea5aa9b42"
                    onChange={[Function]}
                    placeholder="Spanish Translation"
                    showLabel={false}
                  />
                </div>
              </div>
              <div
                className="_translating_operator_container_c108c5"
                data-spec="operator-field"
              >
                <div
                  className="_translating_operator_c108c5"
                  data-spec="verbose-name"
                >
                  has any of the words
                </div>
                <div
                  className="_translating_from_c108c5"
                  data-spec="argument-to-translate"
                >
                  blue
                </div>
                <div
                  className="_translating_to_c108c5"
                >
                  <TextInputElement
                    data-spec="localize-case"
                    entry={
                      {
                        "value": "",
                      }
                    }
                    name="763e4844-3e1b-407a-a1b5-5fdfcd308b41"
                    onChange={[Function]}
                    placeholder="Spanish Translation"
                    showLabel={false}
                  />
                </div>
              </div>
            </div>
          </div>
        </React.Fragment>,
        "name": "Rule Translations",
      },
    ]
  }
  title="Spanish Category Names"
>
  <p
    data-spec="instructions"
  >
    When category names are referenced later in the flow, the appropriate language for the category will be used. If no translation is provided, the original text will be used.
  </p>
  <div
    className="_translating_list_container_c108c5 "
    tabIndex={0}
  >
    <div
      className="_translating_item_list_c108c5"
    >
      <div
        className="_translating_category_c108c5"
        key="1e47a1e1-3c67-4df5-adf1-da542c789adb"
      >
        <div
          className="_translating_from_c108c5"
          data-spec="category-name"
        >
          Red
        </div>
        <div
          className="_translating_to_c108c5"
        >
          <TextInputElement
            data-spec="localize-category"
            entry={
              {
                "value": "",
              }
            }
            name=""
            onChange={[Function]}
            placeholder="Spanish Translation"
            showLabel={false}
          />
        </div>
      </div>
      <div
        className="_translating_category_c108c5"
        key="2dc85899-0577-4f1b-a620-f12094e34b5e"
      >
        <div
          className="_translating_from_c108c5"
          data-spec="category-name"
        >
          Blue
        </div>
        <div
          className="_translating_to_c108c5"
        >
          <TextInputElement
            data-spec="localize-category"
            entry={
              {
                "value": "",
              }
            }
            name=""
            onChange={[Function]}
            placeholder="Spanish Translation"
            showLabel={false}
          />
        </div>
      </div>
      <div
        className="_translating_category_c108c5"
        key="70ac6ea6-803a-4c33-81c7-dc26803c313f"
      >
        <div
          className="_translating_from_c108c5"
          data-spec="category-name"
        >
          Other
        </div>
        <div
          className="_translating_to_c108c5"
        >
          <TextInputElement
            data-spec="localize-category"
            entry={
              {
                "value": "Otro",
              }
            }
            name="Otro"
            onChange={[Function]}
            placeholder="Spanish Translation"
            showLabel={false}
          />
        </div>
      </div>
    </div>
  </div>
</Dialog>
`;

exports[`RouterLocalizationForm > updates > should save changes 1`] = `
{
  "cases": [
    {
      "arguments": [
        "Red, r, rolo, maroon",
      ],
      "category_uuid": "1e47a1e1-3c67-4df5-adf1-da542c789adb",
      "type": "has_any_word",
      "uuid": "061fc171-8b79-4636-b892-bd0ea5aa9b42",
    },
    {
      "uuid": "763e4844-3e1b-407a-a1b5-5fdfcd308b41",
    },
  ],
  "categories": [
    {
      "exit_uuid": "477ac8b4-25e2-483a-8686-2d1332c4da1c",
      "name": "Roooojo!",
      "uuid": "1e47a1e1-3c67-4df5-adf1-da542c789adb",
    },
    {
      "name": "",
      "uuid": "2dc85899-0577-4f1b-a620-f12094e34b5e",
    },
    {
      "exit_uuid": "42ebf1d0-a43e-4ba6-9577-c863b20c88c4",
      "name": "Otro",
      "uuid": "70ac6ea6-803a-4c33-81c7-dc26803c313f",
    },
  ],
  "valid": true,
}
`;

exports[`RouterLocalizationForm > updates > should save changes 2`] = `
[MockFunction spy] {
  "calls": [
    [
      "spa",
      [
        {
          "translations": {
            "name": "Roooojo!",
          },
          "uuid": "1e47a1e1-3c67-4df5-adf1-da542c789adb",
        },
        {
          "uuid": "2dc85899-0577-4f1b-a620-f12094e34b5e",
        },
        {
          "translations": {
            "name": "Otro",
          },
          "uuid": "70ac6ea6-803a-4c33-81c7-dc26803c313f",
        },
        {
          "translations": {
            "arguments": [
              "Red, r, rolo, maroon",
            ],
          },
          "uuid": "061fc171-8b79-4636-b892-bd0ea5aa9b42",
        },
        {
          "uuid": "763e4844-3e1b-407a-a1b5-5fdfcd308b41",
        },
      ],
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
